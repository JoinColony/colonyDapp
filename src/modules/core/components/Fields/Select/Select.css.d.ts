export const baseSelect: string;
export const main: string;
export const containerHorizontal: string;
export const unselectableHorizontal: string;
export const withoutPadding: string;
export const selectInner: string;
export const unselectable: string;
export const inputWrapper: string;
export const select: string;
export const selectExpandContainer: string;
export const activeOption: string;
export const themeDefault: string;
export const themeAlt: string;
export const themeGrey: string;
export const themeGrid: string;
export const widthContent: string;
export const widthFluid: string;
export const widthStrict: string;
export const sizeSmall: string;
export const sizeMedium: string;
export const sizeMediumLarge: string;
export const sizeLarge: string;
export const activeOptionColorBlue: string;
export const activeOptionColorHighlighted: string;
