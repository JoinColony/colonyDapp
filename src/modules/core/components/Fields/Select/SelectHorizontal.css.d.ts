export const horizontalWrapper: string;
export const status: string;
export const unselectable: string;
export const withoutPadding: string;
