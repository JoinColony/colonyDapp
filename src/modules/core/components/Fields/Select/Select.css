.baseSelect {
  padding: 5px 10px;
  width: auto;
  width: 100%;
  position: relative;
  border: 0;
  font-size: var(--size-normal);
  font-style: normal;
  color: var(--text);
  transition: none;
  cursor: pointer;
  outline-color: var(--primary);
  outline-width: 1px;
}

.main {
  composes: field from '~styles/fields.css';
}

.selectInner {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.inputWrapper {
  position: relative;
}

.select[aria-disabled="true"], .select[aria-busy="true"] {
  opacity: 0.7;
}

.select svg {
  fill: var(--grey-4);
  stroke: var(--grey-4);
}

.select[aria-expanded="true"] i {
  transform: rotate(180deg);
}

.selectExpandContainer {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 21px;
  width: 21px;
}

.activeOption {
  height: 100%;
  overflow: hidden;
  font-weight: var(--weight-bold);
  text-overflow: ellipsis;
  white-space: nowrap;
}

.select[aria-expanded="true"] .activeOption, .select[aria-expanded="true"]:focus {
  color: var(--sky-blue);
}

/* themes */
.themeDefault {
  composes: baseSelect;
  border: 1px solid var(--temp-grey-5);
  background-color: white;
}

.themeDefault:focus {
  border-color: var(--primary);
  outline: 0;
}

.themeAlt {
  composes: baseSelect;
  padding: 5px 0 5px 5px;
  border: none;
  border-radius: var(--radius-normal);
  background-color: transparent;
  font-size: var(--size-small);
  outline: none;
}

.themeAlt.select[aria-expanded="true"], .themeAlt:focus {
  background-color: var(--temp-grey-5);
}

.themeGrey {
  composes: baseSelect;
  padding: 10.5px 5px 10.5px 15px;
  border: 1px solid rgb(200, 214, 245);
  border-radius: var(--radius-tiny);
  background-color: rgb(249, 250, 250);
  color: rgb(60, 68, 77);
  box-shadow: inset 0px 2px 4px rgba(0, 0, 0, 0.07);
}

.themeGrey .activeOption {
  font-weight: var(--weight-normal);
}

.widthContent {
  width: min-content;
}

.widthFluid {
  width: 100%;
}

.widthStrict {
  width: 100px;
}
