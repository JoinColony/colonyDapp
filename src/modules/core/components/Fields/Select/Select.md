
### Basic Select Input

```jsx
import { Formik } from 'formik';
import Button from '../../Button';

const options = [
  { label: 'Option 1', value: 1 },
  { label: 'Option 2', value: 2 },
  { label: 'Option three', value: 'three' },
];
<Formik
  onSubmit={(values) => console.log(values)}
  render={({ handleSubmit }) => (
    <form onSubmit={handleSubmit}>
      <Select
        label="I'm a Select"
        options={options}
        placeholder="Select an option"
        name="basicSelect"
      />
      <Select
        appearance={{ alignOptions: 'left', theme: 'alt', width: 'fluid' }}
        label="I'm an alt Select"
        options={options}
        placeholder="Select an alt option"
        name="selectAltTheme"
      />
      <Button type="submit">Press Me</Button>
    </form>
  )}
/>
```

### Unconnected select input

```jsx
import Button from '../../Button';

const options = [
  { label: 'Option 1', value: 1 },
  { label: 'Option 2', value: 2 },
  { label: 'Option three', value: 'three' },
];
initialState = { $value: ''};
<div style={{width: '120px'}}>
  <Select
    connect={false}
    // with connect={false}, `$value` and `setValue` are required
    $value={state.$value}
    setValue={val => setState({ $value: val })}
    appearance={{ alignOptions: 'right', theme: 'alt', width: 'strict' }}
    elementOnly={true}
    label="I'm an unconnected Select"
    options={options}
    placeholder="Select"
    name="selectUnconnected"
  />
  <Button
    onClick={() => {
      console.log(`Currenly chosen value: ${state.$value}`);
    }}
  >
    Press Me
  </Button>
</div>
```
