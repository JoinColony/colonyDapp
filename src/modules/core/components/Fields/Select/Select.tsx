import React, {
  KeyboardEvent,
  ReactNode,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { defineMessages, MessageDescriptor, useIntl } from 'react-intl';
import { useField } from 'formik';
import { nanoid } from 'nanoid';
import classNames from 'classnames';

import { getMainClasses } from '~utils/css';
import { DOWN, ENTER, ESC, SimpleMessageValues, SPACE, UP } from '~types/index';

import SelectListBox from './SelectListBox';
import { Appearance, SelectOption } from './types';
import InputLabel from '../InputLabel';
import InputStatus from '../InputStatus';
import Icon from '../../Icon';

import styles from './Select.css';

const MSG = defineMessages({
  expandIconHTMLTitle: {
    id: 'Select.expandIconHTMLTitle',
    defaultMessage: 'expand',
  },
});

export interface Props {
  /** Appearance object */
  appearance?: Appearance;

  /** Should `select` be disabled */
  disabled?: boolean;

  /** Should render the select without a label */
  elementOnly?: boolean;

  /** Help text */
  help?: string | MessageDescriptor;

  /** Help text values for intl interpolation */
  helpValues?: SimpleMessageValues;

  /** Html id attribute */
  id?: string;

  /** Label text */
  label: string | MessageDescriptor;

  /** Label text values for intl interpolation */
  labelValues?: SimpleMessageValues;

  /** Html name attribute */
  name: string;

  /** Callback function, called after value is changed */
  onChange?: (value: SelectOption['value']) => void;

  /** Available `option`s for the select */
  options: SelectOption[];

  /** Render at the bottom of the select list box */
  optionsFooter?: ReactNode;

  /** Status text */
  placeholder?: string | MessageDescriptor;

  /** Render the actively selected option */
  renderActiveOption?: (
    activeOption: SelectOption | undefined,
    activeOptionLabel: string,
  ) => ReactNode;

  /** Status text */
  status?: string | MessageDescriptor;

  /** Status text values for intl interpolation */
  statusValues?: SimpleMessageValues;

  /** Provides value for data-test prop in select button used on cypress testing */
  dataTest?: string;

  /** Provides value for data-test prop in select items used on cypress testing */
  itemDataTest?: string;

  unselectable?: boolean;
}

const displayName = 'Select';

const Select = ({
  appearance,
  disabled,
  elementOnly,
  help,
  helpValues,
  id: idProp,
  label,
  labelValues,
  name,
  onChange: onChangeCallback,
  options,
  optionsFooter,
  placeholder,
  renderActiveOption,
  status,
  statusValues,
  dataTest,
  itemDataTest,
  unselectable,
}: Props) => {
  const [id] = useState<string>(idProp || nanoid());
  const [, { error, value }, { setValue }] = useField(name);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const { formatMessage } = useIntl();

  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [selectedOption, setSelectedOption] = useState<number>(-1);

  const checkedOption = useMemo(
    () => options.findIndex((option) => option.value === value),
    [value, options],
  );

  const open = () => {
    if (disabled) return;
    setIsOpen(true);
  };

  const close = useCallback(() => {
    setIsOpen(false);
    setSelectedOption(-1);
  }, []);

  const handleOutsideClick = useCallback(
    (evt: MouseEvent) => {
      if (
        wrapperRef.current &&
        evt.target instanceof Node &&
        !wrapperRef.current.contains(evt.target)
      ) {
        close();
      }
    },
    [close],
  );

  const goUp = () => {
    if (selectedOption > 0) {
      setSelectedOption(selectedOption - 1);
    }
  };

  const goDown = () => {
    if (selectedOption < options.length - 1) {
      setSelectedOption(selectedOption + 1);
    }
  };

  const checkOption = () => {
    if (selectedOption === checkedOption || selectedOption === -1) {
      // No change
      close();
      return;
    }
    const { value: optionValue } = options[selectedOption];
    setValue(optionValue);
    if (onChangeCallback) {
      onChangeCallback(optionValue);
    }
    close();
  };

  const handleKeyOnOpen = (evt: KeyboardEvent<any>) => {
    const { key } = evt;
    switch (key) {
      case SPACE: {
        // prevent page long-scroll when in view
        evt.preventDefault();
        close();
        break;
      }
      case UP: {
        // prevent page scroll when in view
        evt.preventDefault();
        goUp();
        break;
      }
      case DOWN: {
        // prevent page scroll when in view
        evt.preventDefault();
        goDown();
        break;
      }
      case ENTER: {
        // Do not submit form
        evt.preventDefault();
        checkOption();
        break;
      }
      default:
    }
  };

  const handleKeyOnClosed = (evt: KeyboardEvent<HTMLElement>) => {
    const { key } = evt;
    if ([UP, DOWN, SPACE].indexOf(key) > -1) {
      evt.preventDefault();
      setSelectedOption(checkedOption);
      open();
    }
  };

  const handleKeyUp = (evt: KeyboardEvent<HTMLElement>) => {
    // Special keyUp handling for ESC (modals)
    const { key } = evt;
    if (isOpen && key === ESC) {
      evt.stopPropagation();
      close();
    }
  };

  const handleKeyDown = (evt: KeyboardEvent<HTMLElement>) => {
    if (isOpen) {
      handleKeyOnOpen(evt);
      return;
    }
    handleKeyOnClosed(evt);
  };

  const toggle = () => {
    if (isOpen) {
      close();
    } else {
      open();
    }
  };

  const selectOption = (idx: number) => {
    setSelectedOption(idx);
  };

  useEffect(() => {
    if (isOpen) {
      if (document.body) {
        document.body.addEventListener('click', handleOutsideClick, true);
      }
    }
    return () => {
      if (document.body) {
        document.body.removeEventListener('click', handleOutsideClick, true);
      }
    };
  }, [handleOutsideClick, isOpen]);

  const activeOptionDisplay = useMemo<ReactNode>(() => {
    /*
     * @NOTE If the active option is removed by something (ie: filtered out),
     * fall back to the last entry in the options array
     */
    const activeOption = options[checkedOption] || options[options.length - 1];
    let activeOptionLabel;
    if (activeOption) {
      if (typeof activeOption.label === 'object') {
        activeOptionLabel = formatMessage(
          activeOption.label,
          activeOption.labelValues,
        );
      } else if (activeOption.children) {
        activeOptionLabel = activeOption.children;
      } else if (activeOption.labelElement) {
        activeOptionLabel = activeOption.labelElement;
      } else {
        activeOptionLabel = activeOption.label;
      }
    }
    const activeOptionLabelText = activeOptionLabel || placeholder;
    if (renderActiveOption) {
      return renderActiveOption(activeOption, activeOptionLabelText);
    }
    return <span>{activeOptionLabelText}</span>;
  }, [checkedOption, formatMessage, options, placeholder, renderActiveOption]);

  const listboxId = `select-listbox-${id}`;

  const containerClasses = classNames(styles.main, {
    [styles.containerHorizontal]: appearance?.direction === 'horizontal',
    [styles.unselectableHorizontal]: !!unselectable,
  });

  const selectOptions = unselectable ? options.slice(0, -1) || [] : options;

  return (
    <div className={containerClasses} ref={wrapperRef}>
      <InputLabel
        inputId={id}
        label={label}
        labelValues={labelValues}
        help={help}
        helpValues={helpValues}
        screenReaderOnly={elementOnly}
        horizontal={appearance?.direction === 'horizontal'}
        appearance={{ colorSchema: appearance?.colorSchema }}
      />
      <div className={styles.inputWrapper}>
        <button
          className={`${styles.select} ${getMainClasses(appearance, styles)}`}
          aria-haspopup="listbox"
          aria-controls={listboxId}
          aria-expanded={isOpen}
          aria-label={
            typeof label === 'object'
              ? formatMessage(label, labelValues)
              : label
          }
          aria-disabled={disabled}
          id={id}
          tabIndex={0}
          onClick={toggle}
          onKeyUp={handleKeyUp}
          onKeyDown={handleKeyDown}
          type="button"
          name={name}
          data-test={dataTest}
        >
          <div
            className={classNames(
              styles.selectInner,
              unselectable && styles.unselectable,
            )}
          >
            <div className={styles.activeOption}>{activeOptionDisplay}</div>
            <span className={styles.selectExpandContainer}>
              <Icon name="caret-down-small" title={MSG.expandIconHTMLTitle} />
            </span>
          </div>
        </button>
        {isOpen && !!options.length && (
          <SelectListBox
            checkedOption={checkedOption}
            selectedOption={selectedOption}
            listboxId={listboxId}
            options={selectOptions}
            optionsFooter={optionsFooter}
            onSelect={selectOption}
            onClick={unselectable ? () => {} : checkOption}
            appearance={appearance}
            name={name}
            dataTest={itemDataTest}
            unselectable={unselectable}
          />
        )}
      </div>
      {!elementOnly && (
        <InputStatus
          appearance={{ theme: 'minimal' }}
          status={status}
          statusValues={statusValues}
          error={error}
        />
      )}
    </div>
  );
};

Select.displayName = displayName;

Select.defaultProps = {
  appearance: { alignOptions: 'left', theme: 'default' } as Appearance,
  options: [],
};

export default Select;
