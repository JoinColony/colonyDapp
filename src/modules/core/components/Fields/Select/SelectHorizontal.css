.horizontalWrapper {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.horizontalWrapper > label {
  margin-bottom: 0;
}

.status {
  margin-right: 30px;
  text-align: right;
}

.unselectable {
  align-items: flex-start;
  justify-content: flex-end;
}

.unselectable label {
  padding-top: 14px;
}

.unselectable ul {
  background-color: transparent;
  box-shadow: none;
}

.unselectable li {
  padding: 0;
}

.withoutPadding {
  padding: 0;
}
