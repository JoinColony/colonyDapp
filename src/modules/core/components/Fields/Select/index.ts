export { default } from './Select';
export { default as SelectListBox } from './SelectListBox';
export { Appearance, SelectOption } from './types';
export { default as SelectHorizontal } from './SelectHorizontal';
