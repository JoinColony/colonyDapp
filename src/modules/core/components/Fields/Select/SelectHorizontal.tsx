import React from 'react';
import classNames from 'classnames';

import { useField } from 'formik';
import Select, { Props as SelectProps } from './Select';
import styles from './SelectHorizontal.css';
import InputStatus from '../InputStatus';
import InputLabel from '../InputLabel';

interface Props extends SelectProps {
  unselectable?: boolean;
  optionSizeLarge?: boolean;
  withDropdownElelment?: boolean;
  scrollContainer?: HTMLElement | null;
  placement?: 'bottom' | 'right';
}

const SelectHorizontal = ({
  unselectable,
  optionSizeLarge,
  withDropdownElelment,
  scrollContainer,
  placement,
  ...selectProps
}: Props) => {
  const {
    status,
    statusValues,
    elementOnly,
    name,
    id,
    label,
    labelValues,
    help,
    helpValues,
  } = selectProps;
  const [, { error }] = useField(name);

  return (
    <>
      <div
        className={classNames(styles.horizontalWrapper, {
          [styles.unselectable]: unselectable,
          [styles.optionSizeLarge]: optionSizeLarge,
        })}
      >
        <InputLabel
          inputId={id}
          label={label}
          labelValues={labelValues}
          help={help}
          helpValues={helpValues}
          screenReaderOnly={elementOnly}
        />
        <Select
          {...{
            id,
            withDropdownElelment,
            scrollContainer,
            placement,
            ...selectProps,
          }}
          elementOnly
          optionSizeLarge
        />
      </div>
      {!elementOnly && (error || status) && (
        <div className={styles.status}>
          <InputStatus
            appearance={{ theme: 'minimal' }}
            status={status}
            statusValues={statusValues}
            error={error}
          />
        </div>
      )}
    </>
  );
};

export default SelectHorizontal;
