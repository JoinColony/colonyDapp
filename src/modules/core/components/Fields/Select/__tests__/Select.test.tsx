/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { Form } from '~core/Fields';
import { mountWithIntl } from '~testutils';

import Select from '../Select';

const DEFAULT_PROPS = {
  id: 'someStaticId',
  label: 'Select me',
  name: 'foo',
  options: [
    {
      value: 'foo_value',
      label: 'Foo',
    },
    {
      value: 'bar_value',
      // Yes, it works with MessageDescriptors as well!
      label: { id: 'Select.Bar', defaultMessage: 'Bar' },
    },
  ],
  connect: false,
  'data-test': 'selectButton',
};

jest.mock('~core/UserPickerWithSearch/hooks', () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe('Select component', () => {
  test('Renders initial component', () => {
    const wrapper = mountWithIntl(
      <Form initialValues={{ foo: undefined }} onSubmit={() => {}}>
        <Select {...DEFAULT_PROPS} />
      </Form>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
