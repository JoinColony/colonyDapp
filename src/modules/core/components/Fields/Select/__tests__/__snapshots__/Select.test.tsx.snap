// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component Renders initial component 1`] = `
<Form.ActionForm
  initialValues={
    Object {
      "foo": undefined,
    }
  }
  onSubmit={[Function]}
>
  <Formik
    initialValues={
      Object {
        "foo": undefined,
      }
    }
    onSubmit={[Function]}
  >
    <Form>
      <form
        action="#"
        onReset={[Function]}
        onSubmit={[Function]}
      >
        <Select
          appearance={
            Object {
              "alignOptions": "left",
              "theme": "default",
            }
          }
          connect={false}
          data-test="selectButton"
          id="someStaticId"
          label="Select me"
          name="foo"
          options={
            Array [
              Object {
                "label": "Foo",
                "value": "foo_value",
              },
              Object {
                "label": Object {
                  "defaultMessage": "Bar",
                  "id": "Select.Bar",
                },
                "value": "bar_value",
              },
            ]
          }
        >
          <div>
            <InputLabel
              inputId="someStaticId"
              label="Select me"
            >
              <label
                className=""
                htmlFor="someStaticId"
                id="someStaticId-label"
              >
                <div>
                  <span>
                    Select me
                  </span>
                </div>
              </label>
            </InputLabel>
            <div>
              <button
                aria-controls="select-listbox-someStaticId"
                aria-expanded={false}
                aria-haspopup="listbox"
                aria-label="Select me"
                className="undefined "
                id="someStaticId"
                name="foo"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                tabIndex={0}
                type="button"
              >
                <div>
                  <div>
                    <span>
                      Bar
                    </span>
                  </div>
                  <Icon
                    name="caret-down-small"
                    title={
                      Object {
                        "defaultMessage": "expand",
                        "id": "Select.expandIconHTMLTitle",
                      }
                    }
                  >
                    <i
                      className=""
                      title="expand"
                    >
                      <svg
                        viewBox="0 0 30 30"
                      >
                        <use
                          xlinkHref="#undefined"
                        />
                      </svg>
                    </i>
                  </Icon>
                </div>
              </button>
            </div>
            <InputStatus
              appearance={
                Object {
                  "theme": "minimal",
                }
              }
            >
              <p
                className=""
              />
            </InputStatus>
          </div>
        </Select>
      </form>
    </Form>
  </Formik>
</Form.ActionForm>
`;
